packages/flutter_tools/lib/src/commands/run.dart

```diff
@@ -504,7 +493,37 @@ class RunCommand extends RunCommandBase {
   String get category => FlutterCommandCategory.project;
 
   List<Device>? devices;
-  var webMode = false;
+  Future<WebDevServerConfig?> getWebDevServerConfig() async {
+    // Only support "web mode" with a single web device due to resident runner
+    // refactoring required otherwise.
+
+    if (featureFlags.isWebEnabled &&
+        devices != null &&
+        devices!.length == 1 &&
+        await devices!.single.targetPlatform == TargetPlatform.web_javascript) {
+      final String? webPortArg = stringArg('web-port');
+      final int? webPort = webPortArg != null ? int.tryParse(webPortArg) : null;
+
+      final WebDevServerConfig fileConfig = await WebDevServerConfig.loadFromFile(
+        fileSystem: globals.fs,
+        logger: globals.logger,
+      );
+
+      final HttpsConfig? httpsConfig = fileConfig.https?.copyWith(
+        certPath: stringArg('web-tls-cert-path'),
+        certKeyPath: stringArg('web-tls-cert-key-path'),
+      );
+
+      final WebDevServerConfig webDevServerConfig = fileConfig.copyWith(
+        host: stringArg('web-hostname'),
+        port: webPort,
+        https: httpsConfig,
+        headers: extractWebHeaders(),
```

Comment:
The headers should be combined:
```suggestion
        headers: <String, String>{...fileConfig.headers, ...?extractWebHeaders()},
```